### PR TITLE
Add smooth_factor option to choropleth, GeoJson and TopoJson

### PR DESCRIPTION
**Pull request by JamesGardiner**
@ocefpaf How does this look to you?

### CODE REVIEW

**Comment by ocefpaf** [comment]
@JamesGardiner I am out of office (in a conference), but this looks good to me. Thanks!

@BibMartin if you have some spare time can you take a quick look at this?

**Comment by JamesGardiner** [comment]
Thanks @ocefpaf. @BibMartin, if you're happy with this, #428 can be closed

**Comment by ocefpaf** [comment]
@JamesGardiner  I am back and not I can finally review this.

Can you add some tests? Like different smooth values and the expected HTML. (See the current `GeoJson` test as an example.)

File: folium/folium.py [hunk 1]
```diff
@@ -578,6 +578,10 @@ def choropleth(self, geo_path=None, geo_str=None, data_out='data.json',
             keyword argument will enable conversion to GeoJSON.
         reset: boolean, default False
             Remove all current geoJSON layers, start with new layer
+        smooth_factor: float, default None
+            How much to simplify the polyline on each zoom level. More means
+            better performance and smoother look, and less means more accurate
+            representation. Leaflet defaults to 1.0.
```
Comment (ocefpaf) [inline folium/folium.py]: This repetition of docs makes me wonder if we should be using some sort of docstring interpolation...

Comment (JamesGardiner) [inline folium/folium.py]: Thanks for the review @ocefpaf. In terms of docs, I agree the repetition isn't great. When you say interpolation, do you mean something like this: https://pypi.python.org/pypi/sphinx-paramlinks/? Alternatively the description could be contained in the `GeoJson` class and `TopoJson`/`choropleth` could both say something like `smooth_factor: See GeoJson class documentation for details`.

Comment (ocefpaf) [inline folium/folium.py]: @JamesGardiner I was just thinking out loud about the docs. Let's not worry about it in this PR.

**Comment by JamesGardiner** [comment]
If your happy with these tests @ocefpaf it can be merged to close #427.

File: tests/test_folium.py [hunk 2]
```diff
@@ -373,31 +373,14 @@ def test_geo_json_str(self):
...
-                """)  # noqa
+                """)
         obj = obj_temp.render(this=geo_json, json=json)
         assert ''.join(obj.split())[:-1] in out
 
         bounds = self.map.get_bounds()
         assert bounds == [[18.948267, -171.742517],
                           [71.285909, -66.979601]], bounds
 
-    def test_geo_json_bad_color(self):
```
Comment (ocefpaf) [inline tests/test_folium.py]: Why did you remove this test?

File: tests/test_folium.py [hunk 3]
```diff
@@ -473,6 +456,24 @@ def test_topo_json(self):
...
+                            topojson='objects.or_counties_geo',
+                            smooth_factor=0.5)
+
+        out = self.map._parent.render()
+
+        # Verify TopoJson
+        topo_json = [val for key, val in self.map._children.items()
+                     if isinstance(val, TopoJson)][0]
+        topojson_str = topo_json._template.module.script(topo_json)
+        assert ''.join(topojson_str.split())[:-1] in ''.join(out.split())
```
Comment (ocefpaf) [inline tests/test_folium.py]: This is fine. Thanks!

Can you also add a notebook example showing the new feature? (That is optional we don't need that to merge.)

**Comment by ocefpaf** [comment]
@JamesGardiner the tests looks good. I just want to get our main dev (@BibMartin ) opinion on the PR and we can merge this. (Note that in folium we are all volunteers with day time jobs, so that might take a while :grin:)

File: tests/test_folium.py (hunk 2 above)
Comment (JamesGardiner) [inline tests/test_folium.py]: My mistake, I'll rectify tonight with a commit.

File: tests/test_folium.py (hunk 3 above)
Comment (JamesGardiner) [inline tests/test_folium.py]: Sure thing; I'll have one I can modify quite easily in any case over the next few weeks.

**Comment by JamesGardiner** [comment]
No problem @ocefpaf, completely understandable. Thanks for taking the time to review!

**Comment by BibMartin** [comment]
It seems good to me ; thanks @JamesGardiner !
I think we can merge when the removed `test_geo_json_bad_color` is back.

Thanks @ocefpaf for pinging me as *main dev* ; I'm not sure to deserve it (based on my poor reactivity theses past weeks :wink: )

**Comment by ocefpaf** [comment]
> Thanks @ocefpaf for pinging me as *main dev* ; I'm not sure to deserve it (based on my poor reactivity theses past weeks :wink: ) 


Compare the me you are still the main dev here :wink: 

@JamesGardiner ping me once you fix the test and we can merge this!

**Comment by JamesGardiner** [comment]
@ocefpaf I've put the removed test back in, but now have a travis failure for Python 3.4 https://travis-ci.org/python-visualization/folium/jobs/132498993. This was a 500 error from binstar; triggering a retest might let it pass. All tests pass locally.

File: tests/test_folium.py [hunk 4]
```diff
@@ -355,7 +355,7 @@ def test_geo_json_str(self):
         for feature in data['features']:
             feature.setdefault('properties', {}).setdefault('style', {}).update({  # noqa
                 'color': 'black',
-                'opactiy': 1,
+                'opacity': 1,
```
Comment (ocefpaf) [inline tests/test_folium.py]: :+1:

**Comment by ocefpaf** [comment]
Thanks for the awesome contribution @JamesGardiner! Looking forward to merge more PRs :wink:

**Comment by JamesGardiner** [comment]
Cheers @ocefpaf. I'm sure there will be 🎉
